File: 宏基因组/Script/alpha_diver_update.R
library(vegan)
library(dplyr)
library(tibble)
library(ggsignif)
library(purrr)
library(plotly)
library(htmlwidgets)
library(ggplot2)

# 允许通过环境变量覆盖 R 库路径
rlib <- Sys.getenv('METAGE_RLIB', '/root/anaconda3/envs/r/lib/R/library')
if (nzchar(rlib) && dir.exists(rlib)) {
    .libPaths(rlib)
}

args <- commandArgs(trailingOnly = TRUE)
stopifnot('请提供 3 个参数：data_dir taxdir res_dir' = length(args) >= 3)
data_dir <- args[1]
taxdir <- args[2]
res_dir <- args[3]
font_family <- Sys.getenv('METAGE_FONT', '宋体')

# 读入样本元数据
sample <- read.table(
    file.path(data_dir, 'sample-metadata.tsv'),
    sep = '\t', header = TRUE, check.names = FALSE, fill = TRUE
)
n_groups <- ncol(sample) - 1

for (g in seq_len(n_groups)) {
    gro_num <- paste0('group', g)
    genecountfile <- file.path(taxdir, gro_num, '4-GeneAbundance', 'gene_count.csv')
    taxfile <- file.path(taxdir, gro_num, '5-TaxAnnotation', '1.Tables', 'gene.taxonomy.csv')

    if (!file.exists(genecountfile)) {
        stop(paste('文件不存在:', genecountfile))
    }
    if (!file.exists(taxfile)) {
        stop(paste('文件不存在:', taxfile))
    }

    otu <- read.csv(genecountfile, header = TRUE, row.names = 1, check.names = FALSE)
    otu$genes <- rownames(otu)
    tax <- read.csv(taxfile, header = TRUE, row.names = 1, check.names = FALSE)
    tax$genes <- rownames(tax)

    tax_otu_raw <- merge(tax, otu, by = 'genes')
    tax_otu_raw <- tax_otu_raw[, 8:ncol(tax_otu_raw)]

    tax_otu_raw <- tax_otu_raw %>%
        group_by(species) %>%
        summarise_all(sum) %>%
        as.data.frame() %>%
        column_to_rownames('species')
    tax_otu_raw <- t(tax_otu_raw)

    samples <- sample %>%
        mutate(across(everything(), as.character))
    samples <- samples[samples[, gro_num] != '', c('sample.id', gro_num)]

    for (j in c('All', 'Archaea', 'bacteria', 'Fungi', 'Virus')) {
        print(j)
        specicestaxfile <- file.path(
            taxdir, gro_num, '5-TaxAnnotation', '1.Tables', 'Samples', j, paste0(j, '.taxonomy.csv')
        )
        if (!file.exists(specicestaxfile)) {
            warning(paste('文件不存在，跳过:', specicestaxfile))
            next
        }
        sp <- read.csv(specicestaxfile, header = TRUE, check.names = FALSE)
        tax_otu <- tax_otu_raw[samples$`sample.id`, colnames(tax_otu_raw) %in% sp$species]
        print(ncol(tax_otu))

        Chao1 <- estimateR(tax_otu)[2, ]
        ACE <- estimateR(tax_otu)[4, ]

        resdir <- file.path(res_dir, gro_num, '5-TaxAnnotation', '7.alpha_diversity_analysis', j)
        if (!file.exists(resdir)) {
            dir.create(resdir, recursive = TRUE)
        }

        Shannon <- diversity(tax_otu, index = 'shannon', base = exp(1))
        Gini_simpson <- diversity(tax_otu, index = 'simpson')

        diver_index <- data.frame(Chao1, ACE, Shannon, Gini_simpson)

        xlsx::write.xlsx(diver_index, file.path(resdir, 'diversity_index.xlsx'))

        diver_index$sample <- rownames(diver_index)
        diver_index$group <- samples[, gro_num]
        diver_index[is.na(diver_index)] <- 0

        calculate_p_value <- function(data1, data2) {
            if (var(data1, na.rm = TRUE) * var(data2, na.rm = TRUE) > 0) {
                if (shapiro.test(data1)$p.value > 0.05 && shapiro.test(data2)$p.value > 0.05) {
                    testtry <- try(t.test(data1, data2), silent = TRUE)
                    if (class(testtry) == 'try-error') {
                        p_value <- NA
                    } else {
                        p_value <- testtry$p.value
                        p_value <- round(p_value, 6)
                    }
                } else {
                    testtry <- try(wilcox.test(data1, data2), silent = TRUE)
                    if (class(testtry) == 'try-error') {
                        p_value <- NA
                    } else {
                        p_value <- testtry$p.value
                        p_value <- round(p_value, 6)
                    }
                }
            } else {
                testtry <- try(wilcox.test(data1, data2), silent = TRUE)
                if (class(testtry) == 'try-error') {
                    p_value <- NA
                } else {
                    p_value <- testtry$p.value
                    p_value <- round(p_value, 6)
                }
            }
            return(p_value)
        }

        for (i in 1:4) {
            index <- colnames(diver_index)[i]
            ANOVA <- aov(diver_index[, i] ~ group, data = diver_index)
            p <- summary(ANOVA)[[1]][['Pr(>F)']][1]
            p <- round(p, 4)
            groups <- unique(diver_index$group)
            group_pairs <- combn(groups, 2, simplify = FALSE)
            p_value_results <- map_df(group_pairs, ~ {
                data1 <- diver_index %>% filter(group == .x[1]) %>% pull(i)
                data2 <- diver_index %>% filter(group == .x[2]) %>% pull(i)
                tibble(
                    group1 = .x[1],
                    group2 = .x[2],
                    p_value = calculate_p_value(data1, data2)
                )
            })
            print(p_value_results)

            # 输出成对比较 p 值表
            if (nrow(p_value_results) > 0) {
                write.table(
                    p_value_results,
                    file = file.path(resdir, paste0(index, '_pairwise_pvalue.tsv')),
                    sep = '\t', row.names = FALSE, quote = FALSE
                )
            }

            if (nrow(p_value_results) > 0) {
                comparisons <- map2(p_value_results$group1, p_value_results$group2, ~ c(.x, .y))
                annotations <- ifelse(
                    p_value_results$p_value < 0.001, '***',
                    ifelse(
                        p_value_results$p_value < 0.01, '**',
                        ifelse(p_value_results$p_value < 0.05, '*', sprintf('p=%.3f', p_value_results$p_value))
                    )
                )
                y_positions <- max(diver_index[, i]) * (1 + 0.1 * seq_along(comparisons))
            } else {
                comparisons <- list()
                annotations <- character()
                y_positions <- numeric()
            }

            diver_p <- ggplot(diver_index, aes(x = group, y = diver_index[, i], color = group)) +
                geom_boxplot() +
                geom_point() +
                {
                    if (length(comparisons) > 0) {
                        geom_signif(
                            comparisons = comparisons,
                            annotations = annotations,
                            y_position = y_positions,
                            tip_length = 0.01,
                            color = 'black'
                        )
                    }
                } +
                theme_bw(base_family = font_family, base_size = 12) +
                theme(
                    panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                    axis.text.x = element_text(angle = 45, hjust = 1),
                    axis.title.x = element_blank(),
                    title = element_text(size = 9)
                ) +
                ggtitle(paste('ANOVA: p=', p)) +
                labs(y = index)

            ggp1 <- ggplotly(diver_p)
            saveWidget(
                ggp1,
                file = file.path(resdir, paste0(index, '.html')),
                selfcontained = TRUE
            )

            ggsave(
                file.path(resdir, paste0(index, '.pdf')),
                diver_p, width = 6, height = 4, device = cairo_pdf
            )
        }
    }
}

File: 宏基因组/Script/alpha_diver_update.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Pure-Python replacement for alpha_diver_update.R.

CLI: python alpha_diver_update.py <data_dir> <taxdir> <res_dir>
"""

import os
import sys
import argparse
import logging
from itertools import combinations

import numpy as np
import pandas as pd
from scipy import stats

import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

import plotly.graph_objects as go
from plotly.subplots import make_subplots

logging.basicConfig(
    level=logging.INFO,
    format='[%(asctime)s] %(levelname)s: %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)
log = logging.getLogger(__name__)

CLASSES = ['All', 'Archaea', 'bacteria', 'Fungi', 'Virus']
INDEX_NAMES = ['Chao1', 'ACE', 'Shannon', 'Gini_simpson']

plt.rcParams['font.sans-serif'] = ['SimHei', 'DejaVu Sans', 'Arial Unicode MS', 'sans-serif']
plt.rcParams['axes.unicode_minus'] = False


def _read_metadata(data_dir):
    path = os.path.join(data_dir, 'sample-metadata.tsv')
    df = pd.read_csv(path, sep='\t', skiprows=[1], dtype=str)
    return df


def _chao1(counts):
    counts = counts[counts > 0]
    s_obs = len(counts)
    f1 = np.sum(counts == 1)
    f2 = np.sum(counts == 2)
    if f2 > 0:
        return s_obs + (f1 ** 2) / (2.0 * f2)
    if f1 > 0:
        return s_obs + f1 * (f1 - 1) / 2.0
    return float(s_obs)


def _ace(counts):
    counts = np.asarray(counts, dtype=float)
    s_abund = np.sum(counts > 10)
    rare = counts[counts > 0]
    rare = rare[rare <= 10]
    s_rare = len(rare)
    n_rare = rare.sum()
    if n_rare == 0 or s_rare == 0:
        return float(np.sum(counts > 0))
    f1 = np.sum(rare == 1)
    c_ace = 1.0 - f1 / n_rare
    if c_ace == 0:
        return float(np.sum(counts > 0))
    k = np.arange(1, 11)
    f_k = np.array([np.sum(rare == i) for i in k])
    gamma2 = max((s_rare / c_ace) * np.sum(k * (k - 1) * f_k) / (n_rare * (n_rare - 1)) - 1.0, 0.0)
    return float(s_abund + s_rare / c_ace + f1 / c_ace * gamma2)


def _shannon(counts):
    total = counts.sum()
    if total == 0:
        return 0.0
    p = counts[counts > 0] / total
    return float(-np.sum(p * np.log(p)))


def _simpson(counts):
    total = counts.sum()
    if total == 0:
        return 0.0
    p = counts / total
    return float(1.0 - np.sum(p * p))


def _alpha_indices(otu):
    """
    otu: DataFrame samples x species (raw counts).
    Returns DataFrame samples x indices.
    """
    try:
        import skbio.diversity.alpha as skalpha
        rows = []
        for idx, row in otu.iterrows():
            counts = row.values.astype(int)
            rows.append({
                'Chao1': skalpha.chao1(counts),
                'ACE': skalpha.ace(counts),
                'Shannon': skalpha.shannon(counts, base=np.e),
                'Gini_simpson': skalpha.simpson(counts)
            })
        return pd.DataFrame(rows, index=otu.index)
    except Exception:
        pass

    rows = []
    for idx, row in otu.iterrows():
        counts = row.values.astype(float)
        rows.append({
            'Chao1': _chao1(counts),
            'ACE': _ace(counts),
            'Shannon': _shannon(counts),
            'Gini_simpson': _simpson(counts)
        })
    return pd.DataFrame(rows, index=otu.index)


def _two_group_pvalue(a, b):
    a = np.asarray(a, dtype=float)
    b = np.asarray(b, dtype=float)
    a = a[~np.isnan(a)]
    b = b[~np.isnan(b)]
    if len(a) < 2 or len(b) < 2:
        return np.nan
    var_a = np.var(a, ddof=1) if len(a) > 1 else 0.0
    var_b = np.var(b, ddof=1) if len(b) > 1 else 0.0
    if len(a) >= 3 and len(b) >= 3 and var_a > 0 and var_b > 0:
        _, pa = stats.shapiro(a)
        _, pb = stats.shapiro(b)
        if pa > 0.05 and pb > 0.05:
            _, p = stats.ttest_ind(a, b, equal_var=False)
            return round(p, 6)
    _, p = stats.ranksums(a, b)
    return round(p, 6)


def _plot_alpha(diver, index, p_anova, pairs, pvals, resdir):
    os.makedirs(resdir, exist_ok=True)
    groups = diver['group'].unique().tolist()
    values = [diver.loc[diver['group'] == g, index].values for g in groups]

    # matplotlib
    fig, ax = plt.subplots(figsize=(6, 4))
    bp = ax.boxplot(values, labels=groups, patch_artist=True)
    for patch, color in zip(bp['boxes'], plt.cm.tab10.colors):
        patch.set_facecolor(color)
    ax.scatter(np.repeat(np.arange(1, len(groups) + 1), [len(v) for v in values]),
               np.concatenate(values), color='black', alpha=0.5, zorder=3)

    if pairs:
        y_max = diver[index].max()
        step = 0.1 * y_max if y_max != 0 else 0.1
        for k, (g1, g2, p) in enumerate(zip(pairs['group1'], pairs['group2'], pvals)):
            i1 = groups.index(g1) + 1
            i2 = groups.index(g2) + 1
            y = y_max + (k + 1) * step
            ax.plot([i1, i1, i2, i2], [y - step / 4, y, y, y - step / 4], color='black', lw=1)
            if p < 0.001:
                txt = '***'
            elif p < 0.01:
                txt = '**'
            elif p < 0.05:
                txt = '*'
            else:
                txt = f'p={p:.3f}'
            ax.text((i1 + i2) / 2.0, y, txt, ha='center', va='bottom', fontsize=9)

    ax.set_title(f'ANOVA: p={p_anova:.4f}')
    ax.set_ylabel(index)
    ax.set_xlabel('')
    plt.setp(ax.get_xticklabels(), rotation=45, ha='right')
    plt.tight_layout()
    fig.savefig(os.path.join(resdir, f'{index}.pdf'), format='pdf', bbox_inches='tight')
    plt.close(fig)

    # plotly
    figp = go.Figure()
    for g, color in zip(groups, plt.cm.tab10.colors):
        figp.add_trace(go.Box(y=diver.loc[diver['group'] == g, index].values, name=g,
                              marker_color=f'rgb{tuple(int(c * 255) for c in color[:3])}',
                              boxpoints='all'))
    if pairs:
        y_max = diver[index].max()
        step = 0.1 * y_max if y_max != 0 else 0.1
        for k, (g1, g2, p) in enumerate(zip(pairs['group1'], pairs['group2'], pvals)):
            i1 = groups.index(g1) + 1
            i2 = groups.index(g2) + 1
            y = y_max + (k + 1) * step
            if p < 0.001:
                txt = '***'
            elif p < 0.01:
                txt = '**'
            elif p < 0.05:
                txt = '*'
            else:
                txt = f'p={p:.3f}'
            figp.add_annotation(x=(i1 + i2) / 2.0, y=y, text=txt, showarrow=False,
                                font=dict(size=12))
    figp.update_layout(title=f'ANOVA: p={p_anova:.4f}', yaxis_title=index, showlegend=False,
                       plot_bgcolor='white', xaxis=dict(showgrid=False))
    figp.write_html(os.path.join(resdir, f'{index}.html'), include_plotlyjs='cdn')


def run_alpha(data_dir, taxdir, res_dir):
    metadata = _read_metadata(data_dir)
    n_groups = metadata.shape[1] - 1

    for g in range(1, n_groups + 1):
        gro_num = f'group{g}'
        genecountfile = os.path.join(taxdir, gro_num, '4-GeneAbundance', 'gene_count.csv')
        taxfile = os.path.join(taxdir, gro_num, '5-TaxAnnotation', '1.Tables', 'gene.taxonomy.csv')
        if not os.path.exists(genecountfile):
            log.error('File not found: %s', genecountfile)
            continue
        if not os.path.exists(taxfile):
            log.error('File not found: %s', taxfile)
            continue

        otu = pd.read_csv(genecountfile, encoding='utf-8-sig', index_col=0)
        tax = pd.read_csv(taxfile, encoding='utf-8-sig', index_col=0)
        otu = otu[~otu.index.duplicated(keep='first')]
        tax = tax[~tax.index.duplicated(keep='first')]

        merged = tax.join(otu, how='inner')
        sample_cols = [c for c in merged.columns if c not in tax.columns]
        tax_otu_raw = merged.groupby('species')[sample_cols].sum().T
        tax_otu_raw.index.name = 'sample'

        group_col = metadata.columns[g]
        samples = metadata[['sample-id', group_col]].copy()
        samples = samples.dropna()
        samples = samples[samples[group_col] != '']
        samples.columns = ['sample-id', 'group']

        for j in CLASSES:
            specicestaxfile = os.path.join(taxdir, gro_num, '5-TaxAnnotation', '1.Tables', 'Samples', j, f'{j}.taxonomy.csv')
            if not os.path.exists(specicestaxfile):
                log.warning('Skip missing taxonomy file: %s', specicestaxfile)
                continue
            sp = pd.read_csv(specicestaxfile, encoding='utf-8-sig')
            keep_species = set(sp['species'].dropna().unique())
            tax_otu = tax_otu_raw[[c for c in tax_otu_raw.columns if c in keep_species]]
            tax_otu = tax_otu.loc[tax_otu.index.isin(samples['sample-id'])]
            if tax_otu.empty:
                log.warning('Empty OTU table for %s %s', gro_num, j)
                continue

            diver_index = _alpha_indices(tax_otu)
            diver_index = diver_index.fillna(0)

            resdir = os.path.join(res_dir, gro_num, '5-TaxAnnotation', '7.alpha_diversity_analysis', j)
            os.makedirs(resdir, exist_ok=True)
            diver_index.to_excel(os.path.join(resdir, 'diversity_index.xlsx'), index=True)

            diver = diver_index.copy()
            diver['sample'] = diver.index
            diver['group'] = samples.set_index('sample-id').loc[diver.index, 'group'].values

            for idx_name in INDEX_NAMES:
                vals = [diver.loc[diver['group'] == gr, idx_name].values for gr in diver['group'].unique()]
                try:
                    fstat, p_anova = stats.f_oneway(*vals)
                except Exception:
                    fstat, p_anova = np.nan, np.nan
                p_anova = round(p_anova, 4) if not np.isnan(p_anova) else np.nan

                groups = diver['group'].unique().tolist()
                pair_records = []
                pvals = []
                for g1, g2 in combinations(groups, 2):
                    a = diver.loc[diver['group'] == g1, idx_name].values
                    b = diver.loc[diver['group'] == g2, idx_name].values
                    p = _two_group_pvalue(a, b)
                    pair_records.append({'group1': g1, 'group2': g2, 'p_value': p})
                    pvals.append(p)
                pair_df = pd.DataFrame(pair_records)
                if not pair_df.empty:
                    pair_df.to_csv(os.path.join(resdir, f'{idx_name}_pairwise_pvalue.tsv'),
                                   sep='\t', index=False)

                _plot_alpha(diver, idx_name, p_anova, pair_df, pvals, resdir)


def main():
    parser = argparse.ArgumentParser(description='Alpha diversity analysis (Python replacement)')
    parser.add_argument('data_dir', help='Directory containing sample-metadata.tsv')
    parser.add_argument('taxdir', help='Taxonomy result directory')
    parser.add_argument('res_dir', help='Output result directory')
    args = parser.parse_args()

    data_dir = os.path.abspath(args.data_dir)
    taxdir = os.path.abspath(args.taxdir)
    res_dir = os.path.abspath(args.res_dir)

    for p in (data_dir, taxdir):
        if not os.path.exists(p):
            log.error('Path does not exist: %s', p)
            sys.exit(1)

    try:
        run_alpha(data_dir, taxdir, res_dir)
        log.info('alpha_diver 完成')
    except Exception as exc:
        log.error('alpha_diver 运行失败: %s', exc)
        sys.exit(1)


if __name__ == '__main__':
    main()
